### PR TITLE
Email notifications

### PR DESCRIPTION
**Pull request by ignaspa**
Still working on testing, mailer working well.

### CODE REVIEW

File: lib/tasks/groups.rake [hunk 1]
```diff
@@ -6,7 +6,7 @@ namespace :db do
     Faker::Config.random = Random.new(42) # seeds the random number generator so Faker output is deterministic
     students = Student.all
     Assignment.all.each do |assignment|
-      num_groups = (assignment.short_identifier == 'A1' || assignment.short_identifier == 'A3') ? students.length : 15
+      num_groups = (assignment.short_identifier == 'A1' || assignment.short_identifier == 'A3') ? 10 : 15
```
Comment (david-yz-liu) [inline lib/tasks/groups.rake]: You should revert the changes in this file (they do not belong in this PR).

File: lib/tasks/users.rake [hunk 2]
```diff
@@ -48,6 +48,9 @@ namespace :db do
       i += rand(10 ** 7)
       student.update_attribute(:id_number, sprintf('%010d', i))
       student.update_attribute(:grace_credits, 5)
+      if student.id == 10
+        student.update_attribute(:email, 'ignaspan@gmail.com')
```
Comment (david-yz-liu) [inline lib/tasks/users.rake]: You should revert this change as well (just for testing purposes).

File: app/helpers/submissions_helper.rb [hunk 3]
```diff
@@ -38,6 +38,14 @@ def set_release_on_results(groupings, release)
       Result.where(id: groupings.joins(:current_result).pluck('results.id'))
             .update_all(released_to_students: release)
     end
+
+    if(release)
+      groupings.each do |grouping|
+        grouping.students.each do |student|
```
Comment (david-yz-liu) [inline app/helpers/submissions_helper.rb]: Do `grouping.accepted_students` here.

File: app/mailers/notification_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+class NotificationMailer < ApplicationMailer
+  def release_email
+    @user = params[:user]
+    @grouping = params[:grouping]
+    attachments.inline['markus_logo_big.png'] = File.read(Rails.root.join('app', 'assets', 'images', 'markus_logo_big.png'))
```
Comment (david-yz-liu) [inline app/mailers/notification_mailer.rb]: I feel like it's better to use a remote URL in the email itself (`asset_url('markus_logo_big.png')` or something) rather than attaching the file. Worried about big emails.

File: app/mailers/notification_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+    mail(to: @user.email, subject: 'MarkUs Notification (' + Rails.configuration.course_name + ') Your marks for ' + @grouping.assignment.short_identifier + ' have released!')
```
Comment (david-yz-liu) [inline app/mailers/notification_mailer.rb]: This is a good start, please internationalize the strings (https://guides.rubyonrails.org/i18n.html). You can add a new subfolder called `config/locales/*/mailers`.

File: app/views/notification_mailer/release_email.html.erb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<div class="emailbkgd dark-scheme">
+  <div class="top_section dark-scheme">
+    <%= image_tag attachments['markus_logo_big.png'].url, :class => "logo_header" %>
+  </div>
+  <hr class="divider">
+  <p class="to_text dark-scheme">
+    Hello <%= @user.first_name %> <%= @user.last_name %>,
```
Comment (david-yz-liu) [inline app/views/notification_mailer/release_email.html.erb]: Similar to above, please internationalize the text in this email..

File: app/views/notification_mailer/release_email.html.erb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  </p>
+  <p class="body_text dark-scheme">
+    Your marks for <%= @grouping.assignment.short_identifier %>
+    have now been released on Markus and can be viewed.
```
Comment (david-yz-liu) [inline app/views/notification_mailer/release_email.html.erb]: And make sure to refer to MarkUs as "Mark**U**s".

File: app/views/notification_mailer/release_email.html.erb [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  </p>
+  <div class="button_enclosure">
+    <%= link_to 'View Assignment Results',
+                view_marks_assignment_submission_result_url(assignment_id: @grouping.assignment.id,
+                                                            submission_id: @grouping.current_submission_used.id,
+                                                            id: @grouping.current_result.id),
+                :style => 'color:#FFFFFF;',
```
Comment (david-yz-liu) [inline app/views/notification_mailer/release_email.html.erb]: This should go into your CSS.

File: app/views/notification_mailer/release_email.html.erb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+                :class => "redirectbutton" %>
```
Comment (david-yz-liu) [inline app/views/notification_mailer/release_email.html.erb]: Use `class: 'redirectbutton'`

File: config/environments/development.rb [hunk 10]
```diff
@@ -242,6 +242,21 @@
   # This variable sets the number of old log files that will be kept
   config.x.logging.old_files = 10
 
+  ###################################################################
+  # Email Notifications
+  ###################################################################
+  config.action_mailer.delivery_method = :smtp
```
Comment (david-yz-liu) [inline config/environments/development.rb]: Make sure all of these new settings are in `test.rb` and `production.rb` as well.

File: lib/tasks/groups.rake (hunk 1 above)
Comment (ignaspa) [reply lib/tasks/groups.rake]: Oh true, I must have commited them by accident!

File: lib/tasks/users.rake (hunk 2 above)
Comment (ignaspa) [reply lib/tasks/users.rake]: Oh, yes and make a PR for emails

File: app/mailers/notification_mailer.rb (hunk 4 above)
Comment (ignaspa) [reply app/mailers/notification_mailer.rb]: Hmm. I think a remote url could work. Part of the reason I did inline was because in development my phone wouldn't be able to access my local host. However, the case stands that if markus crashes then the remote url fails. Unless I make the remote URL point to the github hosting of the image. Also, it should be noted the image is 13.8kB (I would think that is small but I guess we're considering bad email clients or poor data connections?)

File: app/mailers/notification_mailer.rb (hunk 5 above)
Comment (ignaspa) [reply app/mailers/notification_mailer.rb]: Woah, I'll need to look into this more.

File: app/mailers/notification_mailer.rb [hunk 11]
```diff
@@ -0,0 +1,8 @@
+class NotificationMailer < ApplicationMailer
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/mailers/notification_mailer.rb]: @david-yz-liu can I ignore this?

File: app/helpers/submissions_helper.rb [hunk 12]
```diff
@@ -38,6 +38,13 @@ def set_release_on_results(groupings, release)
       Result.where(id: groupings.joins(:current_result).pluck('results.id'))
             .update_all(released_to_students: release)
     end
+    return if(not release)
+      groupings.each do |grouping|
+        grouping.accepted_students.each do |student|
+          NotificationMailer.with(user: student, grouping: grouping).release_email.deliver_now
+        end
+      end
+    end
```
Comment (david-yz-liu) [inline app/helpers/submissions_helper.rb]: And you should remove this `end` as well. Basically, `return if ___` or `return unless ___` isn't a block, but a single statement. You can read more about these statements at https://www.natashatherobot.com/ruby-shorthands-if-then-else/.

File: app/views/notification_mailer/release_email.html.erb [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+                view_marks_assignment_submission_result_url(assignment_id: @grouping.assignment.id,
+                                                            submission_id: @grouping.current_submission_used.id,
+                                                            id: @grouping.current_result.id),
+                class: 'redirectbutton' %>
+  </div>
+  <p class="alert_text dark-scheme">
+    <%= t('disclaimer') %>
+  </p>
+</div>
+
```
Comment (david-yz-liu) [inline app/views/notification_mailer/release_email.html.erb]: Delete the extra blank lines at the end of this file (you should end with only one blank line).

File: spec/mailers/notification_mailer_spec.rb [hunk 14]
```diff
@@ -0,0 +1,31 @@
+require 'rails_helper'
```
Comment (david-yz-liu) [inline spec/mailers/notification_mailer_spec.rb]: You shouldn't need to include this (I believe it's already included in the `spec_helper.rb` file).

File: spec/mailers/notification_mailer_spec.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+RSpec.describe NotificationMailer, type: :mailer do
+  describe 'release_email' do
+    let(:user) { mock_model User, first_name: 'Ignas', last_name: 'Panero Armoska', email: 'ignaspan@gmail.com' }
```
Comment (david-yz-liu) [inline spec/mailers/notification_mailer_spec.rb]: We're actually using [FactoryBot](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md) to create mock objects. Something like `let(:user) { create(:user) }` should work fine.

Comment (david-yz-liu) [reply spec/mailers/notification_mailer_spec.rb]: (And similarly below as well.)

File: spec/mailers/notification_mailer_spec.rb [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+    let(:fake_assignment) { mock_model Assignment, short_identifier: 'A2'}
+    let(:grouping) { mock_model Grouping, assignment: fake_assignment}
+    let(:mail) { described_class.with(user: user, grouping: grouping).release_email.deliver_now }
+
+    it 'renders the subject' do
+      expect(mail.subject).to eq('MarkUs Notification NOT SURE FOR COURSE Your marks for A2 have released!')
```
Comment (david-yz-liu) [inline spec/mailers/notification_mailer_spec.rb]: `Rails.configuration` should still be accessible here.

Also instead of `A2` it's better to do `fake_assignment.short_identifier`.

File: spec/mailers/notification_mailer_spec.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    # not sure what these do
+    it 'assigns @name' do
+      expect(mail.body.encoded).to match(user.name)
+    end
+
+    it 'assigns @confirmation_url' do
+      expect(mail.body.encoded)
+          .to match("http://aplication_url/#{user.id}/confirmation")
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/mailers/notification_mailer_spec.rb]: I think you need to update the URL here.

File: app/mailers/notification_mailer.rb (hunk 4 above)
Comment (david-yz-liu) [reply app/mailers/notification_mailer.rb]: @ignaspa forgot to reply to this in my review.

Ah, you're right that the attachment is small, but I'd still prefer having it remote to keep the email as small as possible. MarkUs might send hundreds of these emails at once. (If MarkUs is down the link won't work either.)

Comment (ignaspa) [reply app/mailers/notification_mailer.rb]: Ok

File: app/helpers/submissions_helper.rb [hunk 18]
```diff
@@ -38,6 +38,13 @@ def set_release_on_results(groupings, release)
       Result.where(id: groupings.joins(:current_result).pluck('results.id'))
             .update_all(released_to_students: release)
     end
+    return if(not release)
```
[the remark this answers is not in the document]
Comment (ignaspa) [reply app/helpers/submissions_helper.rb]: I see

File: spec/controllers/submissions_controller_spec.rb [hunk 19]
```diff
@@ -396,12 +396,28 @@
...
+          is_expected.to respond_with(:success)
+        end
+
+        it 'should send an email to every student whose submission was released' do
+          allow(Assignment).to receive(:find) { @assignment }
+          total_students = 0
+          @assignment.groupings.each do |grouping|
+            total_students += grouping.accepted_students.size
+          end
+          puts "TOTAL STUDENTS = " + total_students.to_s
```
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply spec/controllers/submissions_controller_spec.rb]: You should delete this line.

File: config/locales/mailers/en.yml [hunk 20]
```diff
@@ -0,0 +1,8 @@
+en:
+  notification_mailer:
+    release_email:
+      subject: 'MarkUs Notification (%{course}) Your marks for %{assignment} have released!'
```
Comment (david-yz-liu) [inline config/locales/mailers/en.yml]: The last part should be changed to "have been released".

File: app/views/notification_mailer/release_email.html.erb [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+    <%= t('.text_body', assignment: @grouping.assignment.short_identifier) %>
+  </p>
+  <div class="button_enclosure">
+    <%= link_to t('button_text'),
+                view_marks_assignment_submission_result_url(assignment_id: @grouping.assignment.id,
+                                                            submission_id: @grouping.current_submission_used.id,
+                                                            id: @grouping.current_result.id),
+                class: 'redirectbutton' %>
+  </div>
+  <p class="alert_text dark-scheme">
```
Comment (david-yz-liu) [inline app/views/notification_mailer/release_email.html.erb]: Ah, I realized one thing I'd like to see here is a message saying "If the above button doesn't work, copy-and-paste this link into your web browser: <url>."
